### PR TITLE
Angular. Forms

### PR DESCRIPTION
**Pull request by aleksander1802**
1. Task: [Link](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/forms.md)
2. Screenshot: 
localhost_4200_admin
3. Deploy: [Link](https://aleksander1802.github.io/angular2023Q4/)
4. Done: 11.11.2023 / Deadline: 14.11.2023
5. Score 100 / 100

# Scoring:

- [x] **Login block** is implemented as reactive form (**+10**)
- [x] Validation rules are applied to the **Login block**. Form submission is possible only if all inputs are valid (**+15**)
- [x] The application indicates which inputs are invalid on the **Login block** with appropriate styles and messages (**+15**)
- [x] **Admin page** is implemented as reactive form (**+10**)
- [x] Validation rules are applied to the **Admin page**. Form submission is possible only if all inputs are valid (**+15**)
- [x] The application indicates which inputs are invalid on the **Admin page** with appropriate styles and messages (**+15**)
- [x] "Tags" sub-form is implemented using `FormArray`. "Add tag" button works correctly (**+15**)
- [x] "Reset" button functionality is fully implemented (**+5**)

### CODE REVIEW

File: src/app/auth/pages/login-page/login-page.component.ts [hunk 1]
```diff
@@ -48,15 +30,29 @@ export class LoginPageComponent implements OnInit {
...
+            this.handleSuccessfulLogin();
         } else {
-            this.isSubmited = false;
+            this.handleFailedLogin();
         }
     }
+
+    handleSuccessfulLogin() {
+        this.form.reset();
+        this.router.navigate(['/youtube']);
```
Comment (ShkrobNastya) [inline src/app/auth/pages/login-page/login-page.component.ts]: navigate should be the last one line (after this.isSubmitted = false;)

File: src/app/auth/pages/login-page/login-page.component.html [hunk 2]
```diff
@@ -4,28 +4,25 @@ <h2 class="login__title">Login</h2>
     <div
         class="login__control"
         [ngClass]="{
-            invalid: form.get('login')?.touched && form.get('login')?.invalid
+            invalid: form.get('email')?.touched && form.get('email')?.invalid
```
Comment (ShkrobNastya) [inline src/app/auth/pages/login-page/login-page.component.html]: here is better to create a getter (if value is used more than 2 times)
```
get email() {
    return this.form.get('email');
  }
```

Comment (ShkrobNastya) [reply src/app/auth/pages/login-page/login-page.component.html]: *please, fix it in the whole file

File: src/app/auth/pages/login-page/login-page.component.html [hunk 3]
```diff
@@ -4,28 +4,25 @@ <h2 class="login__title">Login</h2>
...
-            autocomplete="login"
+            id="email"
+            type="email"
+            formControlName="email"
+            autocomplete="email"
         />
         <div
             class="validation"
-            *ngIf="form.get('login')?.touched && form.get('login')?.invalid"
+            *ngIf="form.get('email')?.touched && form.get('email')?.invalid"
```
Comment (ShkrobNastya) [inline src/app/auth/pages/login-page/login-page.component.html]: here is better to create a getter

File: src/app/auth/pages/login-page/login-page.component.ts [hunk 4]
```diff
@@ -48,15 +30,29 @@ export class LoginPageComponent implements OnInit {
...
+        this.isSubmitted = false;
+    }
+
+    handleFailedLogin() {
+        this.isSubmitted = false;
+    }
+
+    getError(controlName: string, errorName: string) {
+        return this.form.get(controlName)?.hasError(errorName);
+    }
```
Comment (ShkrobNastya) [inline src/app/auth/pages/login-page/login-page.component.ts:65]: nice!

File: src/app/core/pages/admin-page/admin-page.component.html [hunk 5]
```diff
@@ -0,0 +1,208 @@
+<div class="admin">
+    <h2 class="admin__title">Create new card</h2>
+
+    <form class="admin__form" [formGroup]="form" (ngSubmit)="onSubmit()">
+        <div
+            class="admin__control"
+            [ngClass]="{
+                invalid:
+                    form.get('title')?.touched && form.get('title')?.invalid
```
Comment (ShkrobNastya) [inline src/app/core/pages/admin-page/admin-page.component.html]: here is better to create a getter

Comment (ShkrobNastya) [reply src/app/core/pages/admin-page/admin-page.component.html]: please, fix it in the whole file

File: src/app/core/pages/admin-page/admin-page.component.html [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            <input
+                id="description"
+                type="text"
+                formControlName="description"
+                autocomplete="description"
+            />
+            <div
+                class="validation"
+                *ngIf="
+                    form.get('description')?.touched &&
```
Comment (ShkrobNastya) [inline src/app/core/pages/admin-page/admin-page.component.html]: here is better to create a getter

File: src/app/core/pages/admin-page/admin-page.component.html [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                    Please enter a link to the image
+                </small>
+            </div>
+        </div>
+
+        <div
+            class="admin__control"
+            [ngClass]="{
+                invalid:
+                    form.get('videoLink')?.touched &&
```
Comment (ShkrobNastya) [inline src/app/core/pages/admin-page/admin-page.component.html]: here is better to create a getter

File: src/app/core/pages/admin-page/admin-page.component.html [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                    Please enter a link to the video
+                </small>
+            </div>
+        </div>
+
+        <div
+            class="admin__control"
+            [ngClass]="{
+                invalid:
+                    form.get('creationDate')?.touched &&
```
Comment (ShkrobNastya) [inline src/app/core/pages/admin-page/admin-page.component.html]: here is better to create a getter

**Review by ShkrobNastya — CHANGES_REQUESTED** [review]
image

I created 2 tags but not able to submit the form (looks like you need to do tag field valid if there is one tag and more)

File: src/app/core/pages/admin-page/admin-page.component.html (hunk 5 above)
Comment (aleksander1802) [reply src/app/core/pages/admin-page/admin-page.component.html]: Fixed.
I made a fix for the entire file. Added getters.

```
    get title() {
        return this.form.get('title');
    }

    get description() {
        return this.form.get('description');
    }

    get imageLink() {
        return this.form.get('imageLink');
    }

    get videoLink() {
        return this.form.get('videoLink');
    }

    get creationDate() {
        return this.form.get('creationDate');
    }

   ```
     [ngClass]="{
                invalid: title?.touched && title?.invalid
     }"

File: src/app/core/pages/admin-page/admin-page.component.html (hunk 6 above)
Comment (aleksander1802) [reply src/app/core/pages/admin-page/admin-page.component.html]: Fixed.
```
*ngIf="description?.touched && description?.invalid"

File: src/app/core/pages/admin-page/admin-page.component.html (hunk 7 above)
Comment (aleksander1802) [reply src/app/core/pages/admin-page/admin-page.component.html]: Fixed.
```
[ngClass]="{
                invalid: videoLink?.touched && videoLink?.invalid
}"

File: src/app/core/pages/admin-page/admin-page.component.html (hunk 8 above)
Comment (aleksander1802) [reply src/app/core/pages/admin-page/admin-page.component.html]: Fixed.
```
[ngClass]="{
                invalid: creationDate?.touched && creationDate?.invalid
}"

File: src/app/auth/pages/login-page/login-page.component.html (hunk 2 above)
Comment (aleksander1802) [reply src/app/auth/pages/login-page/login-page.component.html]: Fixed.
```
    get email() {
        return this.form.get('email');
    }

    get password() {
        return this.form.get('password');
    }


```
```
[ngClass]="{ invalid: email?.touched && email?.invalid }"

File: src/app/auth/pages/login-page/login-page.component.ts (hunk 1 above)
Comment (aleksander1802) [reply src/app/auth/pages/login-page/login-page.component.ts]: Fixed.
```
handleSuccessfulLogin() {
        this.form.reset();
        this.isSubmitted = false;
        this.router.navigate(['/youtube']);
 }
```

File: src/app/auth/pages/login-page/login-page.component.html (hunk 3 above)
Comment (aleksander1802) [reply src/app/auth/pages/login-page/login-page.component.html]: Fixed.
```

get email() {
        return this.form.get('email');
    }

    get password() {
        return this.form.get('password');
    }

```
```
<div class="validation" *ngIf="email?.touched && email?.invalid">


 <div class="validation" *ngIf="password?.touched && password?.invalid">
 ```

**Comment by aleksander1802** [comment]
> image
> 
> I created 2 tags but not able to submit the form (looks like you need to do tag field valid if there is one tag and more)

Of course, I can fix it this way without any problems:
```
createTagFormControl(): FormGroup {
        if (this.tags.length > 1) {
            return this.fb.group({
                tag: [''],
            });
        }

        return this.fb.group({
            tag: ['', Validators.required],
        });
    }
```
But according to the requirements:
- each tag input should be required

Therefore, please allow me not to make corrections.
(My English is bad. So sorry about that.)

**Review by ShkrobNastya — APPROVED** [review]

**Pull request merged**